### DISCUSSION TITLE
GDExtension call functions without hash

### DISCUSSION THREAD

**Discussion by ParadoxV5** [godotengine/godot-proposals#8092] [Scripting]
GDExtension `classdb_get_method_bind` and `variant_get_ptr_utility_function` both require hash values from the JSON. This is meant to guard any forward incompatibility when upgrading Godot version.

Howëver, `object_method_bind_call` has a `GDExtensionCallError *r_error` return arg that (persumably) knows how to report wrong arg counts or types, so the hash check is redundant here. (It’s good for `object_method_bind_ptrcall` tho.)

OTOH, `GDExtensionPtrUtilityFunction` takes an `argc`-`argv` pair for args, similar to “regular call”; but doesn’t return a `GDExtensionCallError *r_error`, similar to “ptrcall”. I much prefer having a `GDExtensionCallError` return for a complete and consistent error handling mechanism.

Language binders [like my WIP](https://github.com/ParadoxV5/Godot.rb) currently had to bundle hashes for *godot knows how many* functions (or at least those of [`@GlobalScope`](https://docs.godotengine.org/en/stable/classes/class_%40globalscope.html), which *rely* on `variant_get_ptr_utility_function`) in our GDExtensions.

(Better yet, instead of returning function pointers, let us call utility functions directly – function name, `argc`, `argv`, `return`, etc. all in one go, similar to how `MethodBind` has `variant_call`.)

**Comment by ParadoxV5** [comment]
> similar to how `MethodBind` has `variant_call`

Edit: not exactly. `variant_call` prefers the script, whereas `classdb_get_method_bind` persumably only accesses the class. A(n extension) script persumably can use `classdb_get_method_bind` to call super, but using `variant_call` will surely (I found out the hard way) cause infinite recursion.

**Comment by dsnopek** [comment]
> Howëver, `object_method_bind_call` has a `GDExtensionCallError *r_error` return arg that (persumably) knows how to report wrong arg counts or types, so the hash check is redundant here.

Unfortunately, when making a ptrcall we don't (and in most cases can't) verify the argument types.

Firstly, we don't check the argument types because ptrcalls aim to be the most performant way to call a native function, so we don't do any checking, leaving it up to the language binding to check before calling. Otherwise, you can end up checking the types twice: once in the binding (for type safe languages) and once in the call.

But, also, arguments in ptrcalls are passed as pointers to the raw value, so `int *`, `float *`, `Object **`, etc, **not** as `Variant`s which can store type information. So, on the receiving end, we can't tell the difference between a `int *` and `Object **` - there's no way to check the type, it's just a random address in memory.

**Reply by ParadoxV5** [reply]
> (It’s good for `object_method_bind_ptrcall` tho.)

**Reply by ParadoxV5** [reply]
(emphasis mine)

> Otherwise, you can end up checking the types twice: once in the binding __(for type safe languages)__ and once in the call.

Indeed. Godot-side type checks therefore aligns more with dynamic type languages like ~~GDScript,~~ Python, JavaScript, Ruby…

**Reply by ParadoxV5** [reply]
Oh, now I see it. `GDExtensionPtrUtilityFunction` also takes raw type pointers as args, making it more align with “ptrcall”. I suppose the `argc` is there because utility functions may be vararg?

Utility functions doesn’t have a “regular call” counterpart (`object_method_bind_call` to `object_method_bind_ptrcall`) for consistency. Well, `MethodBind`’s got a way to call static methods right?
